sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,916 @@
+            // has no indication of whether there might be more items in
+            // the stream! (The stream adapter in tokio-stream simply calls
+            // `Result::transpose()`, so in theory an error is not the end of
+            // the stream.)
+            //
+            // For lack of any direction we stop reading entries from the stream
+            // on the first error. That way we at least don't get stuck retrying
+            // an operation that will always fail.
+            loop {
+                match read_dir.next_entry().await {
+                    Ok(Some(entry)) => {
+                        if let Ok(file_name) = entry.file_name().into_string() {
+                            if let Ok(hash) = ArtifactHash::from_str(&file_name)
+                            {
+                                *map.entry(hash).or_default() += 1;
+                            }
+                        }
+                    }
+                    Ok(None) => break,
+                    Err(err) => {
+                        error!(
+                            &self.log,
+                            "Failed to read dir";
+                            "error" => &err,
+                            "path" => mountpoint.as_str(),
+                        );
+                        break;
+                    }
+                }
+            }
+        }
+        if any_datasets {
+            Ok(map)
+        } else {
+            Err(Error::NoUpdateDataset)
+        }
+    }
+
+    /// Common implementation for all artifact write operations that creates
+    /// a temporary file on all datasets.
+    ///
+    /// Errors are logged and ignored unless a temporary file already exists
+    /// (another task is writing to this artifact) or no temporary files could
+    /// be created.
+    async fn writer(
+        &self,
+        sha256: ArtifactHash,
+    ) -> Result<ArtifactWriter, Error> {
+        let mut inner = Vec::new();
+        let mut last_error = None;
+        for mountpoint in self.storage.artifact_storage_paths().await? {
+            let temp_dir = mountpoint.join(TEMP_SUBDIR);
+            if let Err(err) = tokio::fs::create_dir(&temp_dir).await {
+                if err.kind() != ErrorKind::AlreadyExists {
+                    log_and_store!(
+                        last_error, &self.log, "create", temp_dir, err
+                    );
+                    continue;
+                }
+            }
+
+            let temp_path =
+                Utf8TempPath::from_path(temp_dir.join(sha256.to_string()));
+            let file = match OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(&temp_path)
+                .await
+            {
+                Ok(file) => file,
+                Err(err) => {
+                    if err.kind() == ErrorKind::AlreadyExists {
+                        return Err(Error::AlreadyInProgress { sha256 });
```

Comment:
If the write gets interrupted, and dropped, presumably we'd unlink the temporary file, so "partial writes" wouldn't be a problem, correct?


On the other hand, if we only wrote to one of the M.2s, would this early exit prevent us from updating the other one? Noticing that this is a returned error, rather than a `continue`.

Reply:
In theory, yes, the `Drop` implementation on the various camino-tempfile structs will unlink that temporary file if any writes are interrupted. (Unless sled agent abruptly stops, but when it's restarted it'll remove all the temporary directories.)

If we take this branch and return an error, the files we've created so far will be dropped and unlinked (as we previously created a `Utf8TempPath` for them and then used that to create a `NamedUtf8TempFile`).